torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -1042,6 +1012,27 @@ std::shared_ptr<JitFuture> TensorPipeAgent::send(
   return futureResponseMessage->jitFuture;
 }
 
+void TensorPipeAgent::handleClientError(
+    ClientPipe& clientPipe,
+    const tensorpipe::Error& error) {
+  // When an error occurs on a pipe all pending operations will be aborted and
+  // all callbacks invoked with error, hence we immediately flush all future
+  // messages belonging to this pipe.
+  decltype(clientPipe.pendingResponseMessage_) pendingMsgs;
+  {
+    std::lock_guard<std::mutex> lock(clientPipe.mutex_);
+    std::swap(clientPipe.pendingResponseMessage_, pendingMsgs);
+    clientPipe.inError_ = true;
```

Comment:
It's already so prior to this change. Looks like this is never set back to `false`. So this means, such errors in the pipe is un-recoverable? If so, it means this should be a hard failure that terminate the entire gang?

Reply:
Yes, errors on pipes are unrecoverable. That's out of simplicity: due to each pipe being made up of multiple individual connections (of multiple types, e.g., TCP + InfiniBand + others) it's very tricky to re-establish consistency among them. It's much easier to just throw all these individual connections away and re-establish them all, by reopening an entirely new pipe.

So it is in principle possible to recover from an error, but this needs to be done explicitly by the agent. It's not too hard to do, and I had opened this issue to track it: https://github.com/pytorch/pytorch/issues/40936. I'm ready to implement it, but so far there hasn't really been demand for it, hence I didn't prioritize it.

Also, a single pipe going down will not collapse the entire gang. Other pipes will keep working. The user-facing effect is that RPC requests to one specific remote node will keep failing, while RPC requests to other workers will succeed.